### PR TITLE
Possible fix for #4485

### LINKED ISSUE #4485: Crash when parsing nullptr
### Description

A few days ago I had a crash, and spent a huge amount of time debugging my code, when I realized that if I call the `json::parse()` function with `nullptr`, the whole shebang go up in flames with the message: "Access violation reading location 0x00000000" in input_adapters.hpp line 449.

### Reproduction steps

It's quite simple, call the `json::parse` function with `nullptr`. Reproducible on
- Ubuntu 20.04.6 LTS with gcc 9.4.0 and json 3.9.1
- Win 11, MSVC Version 19.41.34120 and debug branch

### Expected vs. actual results

Based on the [documentations](https://json.nlohmann.me/features/parsing/parse_exceptions/), I would expect an exception (unless someone explicitly turns those off, but it is not the case here).

In reality, we receive a SIGSEV, or Access violation depending on the platform.

See:
```
input_adapters.hpp
line 449

...
contiguous_bytes_input_adapter input_adapter(CharT b)
{
    auto length = std::strlen(reinterpret_cast<const char*>(b));
```
> Exception has occurred: W32/0xC0000005
Unhandled exception at 0x690392C0 (ucrtbased.dll) in test.exe.exe: 0xC0000005: Access violation reading location 0x00000000.

### Minimal code example

```Shell
#include "nlohmann/json.hpp"
#include <iostream>

int main(int argc, char* argv[]) {
    const char * string = nullptr;
    try {
        auto result = nlohmann::json::parse(string);
    } catch(...) {
        std::cout << "Exception caught, no harm." << std::endl;
    }

    std::cout << "All good, nothing to see." << std::endl;
    return 0;
}
```

### Error messages

```Shell
Exception has occurred: W32/0xC0000005
Unhandled exception at 0x690392C0 (ucrtbased.dll) in test.exe.exe: 0xC0000005: Access violation reading location 0x00000000.
```

### Compiler and operating system

Ubuntu 20.04.6 LTS with gcc 9.4.0

### Library version

3.9.1

### Validation

- [X] The bug also occurs if the latest version from the [`develop`](https://github.com/nlohmann/json/tree/develop) branch is used.
- [ ] I can successfully [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests).

### PR DESCRIPTION
**Pull request by jordan-hoang**
**Description**

This fixes an issue with getting a crash when passing in a null string to json::parse(). It instead throws an exception.
Also added in a testcase to check for this.

Link to issue.
https://github.com/nlohmann/json/issues/4485

* * *

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [x]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [x]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header files `single_include/nlohmann/json.hpp` and `single_include/nlohmann/json_fwd.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

### CODE REVIEW

File: tests/src/unit-concepts.cpp [hunk 1]
```diff
@@ -148,3 +147,14 @@ TEST_CASE("concepts")
         }
     }
 }
+
+// This probably shouldn't go here..... should be in some other test file, since this has nothing to do with concepts.
```
Comment (nlohmann) [inline tests/src/unit-concepts.cpp]: Yes, this should go to tests/src/unit-deserialization.cpp.

File: tests/src/unit-concepts.cpp [hunk 2]
```diff
@@ -148,3 +147,14 @@ TEST_CASE("concepts")
...
+TEST_CASE("jh-testing")
+{
+    SECTION("parsing a nullstring")
+    {
+        const char * string = nullptr;
+        CHECK_THROWS_AS(nlohmann::json::parse(string), std::exception);
```
Comment (nlohmann) [inline tests/src/unit-concepts.cpp]: Please use `CHECK_THROWS_WITH_AS` just as the other exception tests.

File: include/nlohmann/detail/input/input_adapters.hpp [hunk 3]
```diff
@@ -446,6 +446,10 @@ template < typename CharT,
                int >::type = 0 >
 contiguous_bytes_input_adapter input_adapter(CharT b)
 {
+    if(b == nullptr)
+    {
+        JSON_THROW(std::invalid_argument("cannot parse a null string"));
```
Comment (nlohmann) [inline include/nlohmann/detail/input/input_adapters.hpp]: Please throw `json::parse_error`. Please check `docs/mkdocs/docs/home/exceptions.md` for an overview.

**Review by nlohmann — CHANGES_REQUESTED** [review]

**Review by nlohmann — CHANGES_REQUESTED** [review]
Please also consider the case that

```cpp
nlohmann::json::parse(nullptr);
```

is called. This is then interpreted as `FILE*`, so please adjust `inline file_input_adapter input_adapter(std::FILE* file)` to also throw and exception if `nullptr` is provided.

**Comment by nlohmann** [comment]
(Also, please update to the latest develop branch to have a green CI)

**Comment by jordan-hoang** [comment]
Gotcha, ahh so that's what it is, I noticed errors going away when I updated / merged from develop.
Thanks for letting me know.

File: include/nlohmann/detail/input/input_adapters.hpp [hunk 4]
```diff
@@ -420,6 +420,10 @@ typename container_input_adapter_factory_impl::container_input_adapter_factory<C
 // Special cases with fast paths
 inline file_input_adapter input_adapter(std::FILE* file)
 {
+    if(file == nullptr)
```
Comment (nlohmann) [inline include/nlohmann/detail/input/input_adapters.hpp]: Please add a test for this.

**Review by nlohmann — CHANGES_REQUESTED** [review]
Please run "make amalgamate" using Astyle 3.1 (https://astyle.sourceforge.net) to fix the CI issues.

**Comment by nlohmann** [comment]
Note "make pretty" is not sufficient as it only indents the files in `include` and `tests`. You must run "make amalgamate" to create the single header files in `single_includes`.

**Comment by jordan-hoang** [comment]
Oh I see, I thought the single_include files weren't ever supposed to be changed. But I realized that meant they weren't supposed to be manually edited, but rather generated with make amalgamate and then pushed.

It's my first time trying out open source so sorry about that.

**Comment by nlohmann** [comment]
> Oh I see, I thought the single_include files weren't ever supposed to be changed. But I realized that meant they weren't supposed to be manually edited, but rather generated with make amalgamate and then pushed.

You are right - the file should not be edited, because any changes would be overwritten once the amalgamation is triggered next time. To avoid this, we have this (rather picky) check in the CI that makes sure the files never differ.

> It's my first time trying out open source so sorry about that.

Please don't worry! Thanks for contributing!

**Comment by nlohmann** [comment]
I will have a look at the remaining errors.

File: tests/src/unit-deserialization.cpp [hunk 5]
```diff
@@ -361,6 +361,10 @@ TEST_CASE("deserialization")
                 "start_object()", "key(one)", "number_unsigned(1)",
                 "end_object()", "parse_error(29)"
             }));
+
+            const char* string = nullptr;
+            CHECK_THROWS_WITH_AS(nlohmann::json::parse(string), "[json.exception.parse_error.101] parse error: attempting to parse an empty input; check that your input string or stream contains the expected JSON", json::parse_error&);
```
Comment (nlohmann) [inline tests/src/unit-deserialization.cpp]:
```suggestion
            json _;
            CHECK_THROWS_WITH_AS(_ = json::parse(string), "[json.exception.parse_error.101] parse error: attempting to parse an empty input; check that your input string or stream contains the expected JSON", json::parse_error&);
```

File: tests/src/unit-deserialization.cpp [hunk 6]
```diff
@@ -361,6 +361,10 @@ TEST_CASE("deserialization")
...
+            CHECK_THROWS_WITH_AS(nlohmann::detail::input_adapter(nullptr), "[json.exception.parse_error.101] parse error: attempting to parse an empty input; check that your input string or stream contains the expected JSON", json::parse_error&);
```
Comment (nlohmann) [inline tests/src/unit-deserialization.cpp]:
```suggestion
            CHECK_THROWS_WITH_AS(_ = json::parse(nullptr), "[json.exception.parse_error.101] parse error: attempting to parse an empty input; check that your input string or stream contains the expected JSON", json::parse_error&);
```

File: include/nlohmann/detail/input/input_adapters.hpp [hunk 7]
```diff
@@ -420,6 +420,10 @@ typename container_input_adapter_factory_impl::container_input_adapter_factory<C
 // Special cases with fast paths
 inline file_input_adapter input_adapter(std::FILE* file)
 {
+    if (file == nullptr)
+    {
+        JSON_THROW(parse_error::create(101, 0, "attempting to parse an empty input; check that your input string or stream contains the expected JSON", nullptr));
```
Comment (nlohmann) [inline include/nlohmann/detail/input/input_adapters.hpp:426]: Add

```cpp
#include <nlohmann/detail/exceptions.hpp>
#include <nlohmann/detail/macro_scope.hpp>
```

to use `parse_error` and `JSON_THROW`.

**Review by nlohmann — CHANGES_REQUESTED** [review]
I added some comments to fix some errors in the CI.

File: tests/src/unit-deserialization.cpp (hunk 5 above)
Comment (nlohmann) [reply tests/src/unit-deserialization.cpp]: (Fixes the warning that the result of `parse` is unused.)

File: tests/src/unit-deserialization.cpp (hunk 6 above)
Comment (nlohmann) [reply tests/src/unit-deserialization.cpp]: (Fixes the warning that the result of `parse` is unused. Furthermore, the user will call `parse` and not create an input adapter.)

File: include/nlohmann/detail/input/input_adapters.hpp [hunk 8]
```diff
@@ -17,6 +17,7 @@
 #include <string> // string, char_traits
 #include <type_traits> // enable_if, is_base_of, is_pointer, is_integral, remove_pointer
 #include <utility> // pair, declval
+#include <nlohmann/detail/exceptions.hpp>
```
Comment (nlohmann) [inline include/nlohmann/detail/input/input_adapters.hpp]: For stylistic reasons, please move this include down right above

```cpp
#include <nlohmann/detail/iterators/iterator_traits.hpp>
```

**Review by nlohmann — CHANGES_REQUESTED** [review]
One minor comment.

**Comment by jordan-hoang** [comment]
Sorry about all these tiny commits, I make the changes on a windows push em and then do the make amalgamate on a linux vm I have. Should really get make setup for windows so I don't have to do that.

**Review by nlohmann — APPROVED** [review]
Looks good to me.

**Comment by nlohmann** [comment]
Thanks a lot!!!

**Pull request merged**